### ISSUE TITLE
Is it possible to make the first slide appear again after the last slide? (Infinite loop) 

### ISSUE THREAD

**Issue by Pgma14** [express-labs/pure-react-carousel#60]
Hello, 

I was wondering if it is possible to have the last slide redirect back to the first slide, so that when a user reaches the last slide on a carousel, they can directly swipe back over to the first slide, without having to back track through all the slides. Right now, a user has to click the "back" navigation button to go back to the first slide, but I am trying to make it so that if they click "next" after the last slide, they are taken back to the first slide. 

Any suggestions are appreciated!

**Comment by Pgma14** [comment]
Any update? This would be an incredibly useful feature!

**Comment by tim-steele** [comment]
@Pgma14 Would probably need to add a "infiniteLoop" property to the carousel that sets the current slide back to 0 instead of disabling the Next button. We could look into it, but also feel free to submit a PR for it.

**Comment by TheFullResolution** [comment]
@Pgma14 before any new feature is added you can also try to use `WithStore` HOC to build custom "next button"... All the necessary information and setState method are exposed in it. I can only imagine that if you display more than one slide at the time it might get tricky.

**Comment by mrbinky3000** [comment]
We would have to clone the first Math.ceil(visible slides / 2) slides and clone Math.ceil(visible slides / 2 ) from the end of the slides and append them to the start and end of the slides respectively. 

When someone reaches the end of the slider, we would instantly disable css-animations, reposition the slider, and then re-enable css-animations.  This gives the illusion of an infinite scroll.

When touching or dragging, we'll have to see if the right x position of the last slide becomes less than  the right of the slider.  We'd also have to see if the left x position of the first slide becomes more than the left x of the slider.  If these happen, we do the same css tricks as stated above.

For added bonus, we could remember the velocity of the drag and re-apply the velocity after we jump to the new slide position in the slider.

Problems:  
- slide index id's for the new slides we cloned.  Perhaps we just generate guid for these.
- Sliders can contain anything.... so cloning can be interesting for complex slide contents.  Especially  things that call redux, or graphql or something like that.
- need to exclude the new slides from the buttons list so we don't end up with total slides + visible slides number of slides.  We should only ever have buttons for "total slides"


In a nutshell, this is something that I have personally done years ago in another slider I built when jquery was all the rage.  Time is the only enemy here.  Time and coffee. There are other changes we at Express are going to have to implement first to meet new requirements as stated in #86

This is definitely on the backlog.  Perhaps I can get the Kanban feature on github up and running?

**Comment by matt-peck** [comment]
I took the time to draft up a demo of what this would like in React if we didn't consider the current implementation. 

Check it out here: https://codepen.io/MattPeck/pen/pZbWjN?editors=0010

The main reason I did this, is because I found that a non-infinite scrolling slider and a carousel (infinite) are fundamentally different. It was necessary to actualy modify the the slide object using a property for order.

Allow the CSS to handle transitions by dynamically translating the slider based on what state we are in.

Then after the transition has completed (currently a manual setTimeout time that matches the transition time, we update the order for each of the items based on the direction we clicked, and the carousel re-renders with the correct text and in the exact same place.

It requires at least 1 slide occur before the first one and one at the end, and it endlesslessy loops through them without adding or removing objects behind the scenes.

In addition theres a little check for the sliding state to ensure the user can't click several times causing some twitching.

I know its not PR worthy for this library, but for those who were wanting to do infinite scrolling, I hope it helps convey at least one way to accomplish this.

**Comment by tim-steele** [comment]
@matt-peck I like it. You may want to change the order of operations to "move card to end" then "slide" otherwise the transition is not smooth when only 4 cards are present: https://codepen.io/timssteele/pen/bjrmBW

This is similar to what @mrbinky3000 was discussing in that the card needs to be added to the end to make a faux infinite scroll.

**Comment by martinmckenna** [comment]
adding my name to the list of people who want this feature

**Comment by janoist1** [comment]
+1

**Comment by Alek997** [comment]
I solved this issue with conditional rendering buttons. In state  i have currentSlideIndex which is starting at 0 , and incremented or decremented in handler functions which is passed to ButtonFirst and ButtonNext in onClick props. When currenSlideIndex is equal to (totalSlides - visibleSlides) than you can switch ButtonNext with ButtonFirst, and on next click on that button, Slider will slide to first item with animation. In ButtonFirst onClick prop, you should set currentSlideIndex back to 0. Same can be done with ButtonBack and ButtonLast.
I know that this solution is hacky, but it worked for me, and maybe it will help someone else.
Also this will trigger render because state is changed but you must rerender whole carousel if you want to change buttons conditionaly.
I hope this can help someone.

**Comment by collegeimprovements** [comment]
What's the current best way to do the infinite loop ?

**Comment by tim-steele** [comment]
@collegeimprovements we haven't developed the feature and there hasn't been a pull request on this. It looks like @Alek997 has a "solution". 

If someone would like to attempt to solve this and open a PR we would be open to it.

**Comment by andersonmendesdev** [comment]
> @collegeimprovements we haven't developed the feature and there hasn't been a pull request on this. It looks like @Alek997 has a "solution".
> 
> If someone would like to attempt to solve this and open a PR we would be open to it.

I believe this is a necessity of any corousel, you implemented autoplay but did not implement the inifite loop which seems to me to be inconsistent, autoplay is meaningless with the carousel resetting the sliders.

**Comment by kmprasad4u** [comment]
Minimal Solution (if you don't need to display First and Last buttons)

```html
<ButtonBack className={styles.btnBack}><Icon iconName={"ChevronLeft"} /></ButtonBack>
<ButtonLast className={styles.btnLast}><Icon iconName={"ChevronLeft"} /></ButtonLast>
<ButtonNext className={styles.btnNext}><Icon iconName={"ChevronRight"} /></ButtonNext>
<ButtonFirst className={styles.btnFirst}><Icon iconName={"ChevronRight"} /></ButtonFirst>
```

```css
.btnLast,
.btnFirst {
    display:none;
}
.btnBack:disabled,
.btnNext:disabled {
    display:none;
}
.btnBack:disabled + .btnLast,
.btnNext:disabled + .btnFirst {
    display:inline-block;
}
```

**Comment by mrbinky3000** [comment]
Just an update, this is being worked on.  We need it at Express for an upcoming iteration of the site.

**Comment by jdconner** [comment]
Is this still being worked on? If so, what's the ETA? If not, I am willing to push up a PR. This is the only feature keeping us from using this lib (https://github.com/express-labs/pure-react-carousel/issues/167)

**Comment by alcoelho** [comment]
+1

it's a great feature

**Comment by tim-steele** [comment]
@jdconner I responded to your comment on another thread. If you have a PR we would welcome it.

**Comment by ccurtin** [comment]
@jdconner just curious, will you be submitting a PR? If not I may attempt to implement this feature myself over the next week or so, but not 100% sure I will have the time for it. May just switch libraries.

**Comment by jdconner** [comment]
@ccurtin I will be submitting a PR when I can. See the other thread (https://github.com/express-labs/pure-react-carousel/issues/167) for the ongoing conversation.

**Comment by jdconner** [comment]
@tim-steele I have the PR ready. Mind giving me permissions to submit a PR?

**Comment by tim-steele** [comment]
@jdconner I published out version 1.23.0 which has your code merged in.

**Comment by jdconner** [comment]
@tim-steele awesome, thank you for the prompt responses! 👍

**Comment by alcoelho** [comment]
@jdconner @tim-steele 

great job, guys!!

**Comment by tim-steele** [comment]
@jdconner and anyone else on this thread - we pushed up the branch `feature/infinite-scroll`. This is our work in progress branch for the infinite scroll where it adds items to end/beginning vs. going back to beginning/end.

This is more of a reference to bring this thread to a close.

**Comment by thejamespower** [comment]
Hey Express, what's the status of `feature/infinite-scroll`? The branch has been untouched for a couple of months now, do you need support on it?

**Comment by tbntdima** [comment]
Hi guys, agreed with @thejamespower lack of this feature prevents me from using this lib in my project 😢

**Comment by pmourer** [comment]
Also had to drop this due to the need (indirectly) for infinite slide. wanted to have a thumbnail carousel below the main carousel and tie the 2 together. Unfortunately, with `visibleSlides > 1` and without infinite slide (duplicating slides after last and vice versa) there is no way to "center" the last however many slides because we have reached "the end" when the last visibleSlide is the last slide in the carousel.. creating a very strange UX. couldn't find a way around this.

**Comment by tim-steele** [comment]
@tbntdima @pmourer we pushed a branch for reference and @jdconner put a pull request in a while back allowing a version of infinite scroll. As has been stated on a couple of threads, if anyone would like to open a PR for this we would accept it.

**Comment by Hagafny** [comment]
Any news on this?

**Comment by drewpoling2** [comment]
Posted a solution for infinite scrolling in one direction [here](https://github.com/express-labs/pure-react-carousel/issues/251#issuecomment-1278577083)
